### ISSUE TITLE
Attribute access doesn't work as a typeguard for unions

### ISSUE THREAD

**Issue by Wizzerinus** [microsoft/pyright#9618]
**Describe the bug**
Mypy and Pyright disagree on the code below: Mypy reveals the type of `x` as `A` and Pyright reveals it as `A | B`.

**Code or Screenshots**

```python
from typing import Literal


class A:
    success: Literal[True]


class B:
    success: Literal[False]


def test(x: A | B):
    if x.success:
        reveal_type(x)
```

**VS Code extension or command-line**
Command line, version 1.1.390.

**Comment by Wizzerinus** [comment]
Another similar example is tuple indexing:

```python
T = tuple[int, str] | tuple[str, int]


def f(t: T):
    if isinstance(t[0], int):
        reveal_type(t[1])
```

```
~/tests
% mypy g.py   
g.py:6: note: Revealed type is "builtins.str"
Success: no issues found in 1 source file

~/tests
% pyright g.py
/home/wizzerinus/tests/g.py
  /home/wizzerinus/tests/g.py:6:21 - information: Type of "t[1]" is "str | int"
0 errors, 0 warnings, 1 information
```

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

See [this issue](https://github.com/microsoft/pyright/issues/6113) for details.

For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).
